Temporal Nexus: Go SDK Feature Guide (#3057)
temporalio/documentation: docs/develop/go/temporal-nexus.mdx

Each [Temporal SDK includes and uses a default Data Converter](https://docs.temporal.io/dataconversion), that encodes payloads in the following order: Null, Byte array, Protobuf JSON, and JSON.
In a polyglot environment, Protobuf and JSON are common choices, but this example uses native Go types.

In this example, there is a service package that describes the Service and Operation names along with input/output types for caller Workflows to use the Nexus Endpoint.

<!--SNIPSTART service-nexus-->
<!--SNIPEND-->

## Develop a Nexus Service & Operation handlers {#develop-nexus-service-operation-handlers}

Nexus Operation handlers are typically defined in the same Worker as the underlying Temporal primitives they abstract. 
Operation handlers can decide if a given Nexus Operation will be synchronous or asynchronous, execute arbitrary code, and invoke underlying Temporal primitives such as a Workflow, Query, Signal, or Update.

The `temporalnexus` package has builders to create Nexus Operations:

- `NewSyncOperation` \- Simple synchronous RPC handlers, such as for Signals
- `NewWorkflowRunOperation` \- Run a Workflow as an asynchronous Nexus Operation

> Start with a `NewSyncOperation` example, and then use `NewWorkflowRunOperation` to start a handler Workflow from a Nexus Operation.

Author: This is great.

```
brew install temporalio/brew/tcld
tcld gen ca --org $YOUR_ORG_NAME --validity-period 1y --ca-cert ca.pem --ca-key ca.key
```

### Create caller and handler Namespaces

Before deploying to Temporal Cloud, it's necessary to create the appropriate Namespaces for both the caller and handler.

```
tcld login

tcld namespace create \
	--namespace <your-caller-namespace> \
	--region us-west-2 \
	--ca-certificate-file 'path/to/your/ca.pem' \
	--retention-days 1

tcld namespace create \
	--namespace <your-target-namespace> \
	--region us-west-2 \
	--ca-certificate-file 'path/to/your/ca.pem' \
	--retention-days 1
```

Alternatively, you can create Namespace through the UI: [https://cloud.temporal.io/Namespaces](https://cloud.temporal.io/Namespaces)

### Create a Nexus Endpoint to route requests from caller to handler

To create a Nexus Endpoint you must have a Developer account role or higher, and have NamespaceAdmin permission on the `--target-namespace`.

```
tcld nexus endpoint create \
  --name <my-nexus-endpoint-name> \
  --target-task-queue my-handler-task-queue \
  --target-namespace <my-target-namespace.account> \

>   --allow-namespace <my-caller-namespace.account> \

Author: We do typically try to use "your" but the go-samples seem to be using "my" so I'll just keep it for now.

  --description-file description.md
```

The `--allow-namespace` is used to build an Endpoint allowlist of caller Namespaces that can use the Nexus Endpoint, as described in Runtime Access Control.

> Alternatively, you can create a Nexus Endpoint through the UI: [https://cloud.temporal.io/nexus](https://cloud.temporal.io/nexus)

Reviewer: I would mention that you can create an endpoint via the UI in the dev server too.

```

### Start a caller Workflow

In order to start the caller Workflow, run the starter.

```go
cd caller

go run ./starter \
	-target-host <your-caller-namespace.account>.tmprl.cloud:7233 \
	-namespace <your-caller-namespace.account> \
	-client-cert 'path/to/your/ca.pem' \
	-client-key 'path/to/your/ca.key'
```

This will result in:

```
2024/07/23 19:57:40 Workflow result: Nexus Echo 👋
2024/07/23 19:57:40 Workflow result: ¡Hola! Nexus 👋
```

### Observability

#### Web UI

A synchronous Nexus Operation will surface in the caller Workflow as follows, with just `NexusOperationScheduled` and `NexusOperationCompleted` events in the caller’s Workflow history:

![Observability in Temporal Cloud Sync](/img/nexus-go/go-sdk-observability-sync.png)

An asynchronous Nexus Operation will surface in the caller Workflow as follows, with `NexusOperationScheduled`, `NexusOperationStarted`, and `NexusOperationCompleted`, in the caller’s Workflow history:

![Observability in Temporal Cloud Async](/img/nexus-go/go-sdk-observability-async.png)

#### Temporal CLI

Nexus pending operations and callbacks are included in the Workflow description:

```
temporal workflow describe -w <ID>

> ```

Reviewer: I would separate this to caller and handler workflows. Handler will have a callback attached while caller will see a pending nexus operation as long as the operation is pending completion.

Prerequisites:

- [Install the latest Temporal CLI](https://docs.temporal.io/develop/go/core-application#run-a-development-server)
- [Install the latest Temporal Go SDK](https://docs.temporal.io/develop/go/core-application#install-a-temporal-sdk) (v1.29.1 or higher)

### Start the Temporal Development Server {#start-the-temporal-development-server}

Start the Temporal Development Server by using the server start-dev command, with `system.enableNexus=true`.
The HTTP port is required for Nexus communications.

```
temporal server start-dev --http-port 7243 --dynamic-config-value system.enableNexus=true
```

This command automatically starts the Web UI, creates the default Namespace, and uses an in-memory database.

The Temporal Server should be available on `localhost:7233` and the Temporal Web UI should be accessible at [http://localhost:8233](http://localhost:8233).

## Create caller and handler Namespaces {#create-caller-handler-namespaces}

Before setting up Nexus endpoints, create separate namespaces for the caller and handler.

```
temporal operator namespace create --namespace my-target-namespace
temporal operator namespace create --namespace my-caller-namespace

> ```

Reviewer 2: Just hopping in here to mention that to align with existing `temporal` CLI docs, we use:

```
docs% temporal operator namespace
Manage Temporal Cluster (Service) Namespaces:

temporal operator namespace [command] [command options]

For example:

temporal operator namespace create \
    --namespace YourNewNamespaceName

Usage:
  temporal operator namespace [command]
```

TMI: `tcld` uses a different convention (specifically `<namespace_id>.<account_id>`, but it is not yet all aligned)

## Create a Nexus Endpoint to route requests from caller to handler {#create-nexus-endpoint}

After establishing caller and handler Namespaces, the next step is to create a Nexus Endpoint to route requests.

```go

> temporal operator nexus endpoint create \

Reviewer: The endpoint name is wrong. I would stick closer to the sample's README and run these steps locally to verify they're correct.

https://github.com/temporalio/samples-go/tree/main/nexus#create-nexus-endpoint

```

## Develop a Nexus Service & Operation handlers {#develop-nexus-service-operation-handlers}

Nexus `Operation` handlers are typically defined in the same Worker as the underlying Temporal primitives they abstract. `Operation` handlers can decide if a given Nexus Operation will be synchronous or asynchronous, execute arbitrary code, and invoke underlying Temporal primitives such as a Workflow, Query, Signal, or Update.

The `temporalnexus` package has builders to create Nexus Operations:

- `NewSyncOperation` \- Simple synchronous RPC handlers, such as for Signals
- `NewWorkflowRunOperation` \- Run a Workflow as an asynchronous Nexus Operation

Start with a `NewSyncOperation` example, and then use `NewWorkflowRunOperation` to start a handler Workflow from a Nexus Operation.

### Develop a Synchronous Nexus Operation handler

The `temporalnexus.NewSyncOperation` builder function is for exposing simple RPC handlers. It’s handler function is provided with an SDK client that can be used for signaling, querying, and listing Workflows, but implementations are free to make arbitrary calls to other services or databases, or perform computations such as this one:

```go

> package handler

Reviewer: Looks like the formatting got broken when you pasted the sample here. Are we using snipsync?

import (
	"go.temporal.io/sdk/temporalnexus"

> 	"github.com/temporalio/samples-go/nexus/service"       ...

Reviewer: Are these three dots meant to be here?

)

var EchoOperation = temporalnexus.NewSyncOperation(  service.EchoOperationName,   func(ctx context.Context, c client.Client, input service.EchoInput, options nexus.StartOperationOptions) (service.EchoOutput, error) {
	return service.EchoOutput(input), nil
})
```

### Develop an Asynchronous Nexus Operation handler to start a Workflow

Use the NewWorkflowRunOperation constructor, which is the easiest way to expose a Workflow as an operation.
See alternatives at [https://pkg.go.dev/go.temporal.io/sdk/temporalnexus](https://pkg.go.dev/go.temporal.io/sdk/temporalnexus).

```go

> var HelloOperation = temporalnexus.NewWorkflowRunOperation(  service.HelloOperationName,   HelloHandlerWorkflow,  func(ctx context.Context, input service.HelloInput, options nexus.StartOperationOptions) (client.StartWorkflowOptions, error) {

Reviewer: Same here, this line is too long, I would try to stick to the sample code.

```go
package caller

import (
	"github.com/temporalio/samples-go/nexus/service"
	"go.temporal.io/sdk/workflow"
)

const (
	TaskQueue    = "my-caller-workflow-task-queue"
	endpointName = "myendpoint"
)

func EchoCallerWorkflow(ctx workflow.Context, message string) (string, error) {
	c := workflow.NewNexusClient(endpointName, service.HelloServiceName)

	fut := c.ExecuteOperation(ctx, service.EchoOperationName, service.EchoInput{Message: message}, workflow.NexusOperationOptions{})

	var res service.EchoOutput
	if err := fut.Get(ctx, &res); err != nil {
		return "", err
	}

	return res.Message, nil
}
```

### Register the caller Workflow in a Worker

After developing the caller Workflow, the next step is to register it with a Worker.

```go
package main

import (
	"log"
	"os"

	"github.com/temporalio/samples-go/nexus/caller"
	"github.com/temporalio/samples-go/nexus/options"

	"go.temporal.io/sdk/client"
	"go.temporal.io/sdk/worker"
)

func main() {
	c, err := client.Dial(client.Options{})
	if err != nil {
		log.Fatalln("Unable to create client", err)
	}
	defer c.Close()

	w := worker.New(c, caller.TaskQueue, worker.Options{})

	w.RegisterWorkflow(caller.EchoCallerWorkflow)

>        w.RegisterWorkflow(caller.HelloCallerWorkflow)

Reviewer: Something happened to the indentation here.

	err = w.Run(worker.InterruptCh())
	if err != nil {
		log.Fatalln("Unable to start worker", err)
	}
}
```

### Develop a starter to start the caller Workflow

To initiate the caller Workflow, a starter program is required.

```go
package main

import (
	"context"
	"log"
	"os"
	"time"

	"go.temporal.io/sdk/client"

	"github.com/temporalio/samples-go/nexus/caller"
	"github.com/temporalio/samples-go/nexus/options"
	"github.com/temporalio/samples-go/nexus/service"
)

func main() {
	c, err := client.Dial(client.Options{})
	if err != nil {
		log.Fatalln("Unable to create client", err)
	}
	defer c.Close()
	runWorkflow(c, caller.EchoCallerWorkflow, "Nexus Echo 👋")
	runWorkflow(c, caller.HelloCallerWorkflow, "Nexus", service.ES)
}
```

> ## Make Nexus calls across Namespaces with a dev Server {#nexus-calls-across-namespaces-dev-server}

Reviewer: Maybe just call this "run the tutorial" or something similar?

With the Workers running, the final step in the local development process is to start a caller Workflow.

Run the starter:

```go 
cd caller
go run ./starter \
    -target-host localhost:7233 \
    -namespace my-caller-namespace
```

This will result in:

```
2024/07/23 19:57:40 Workflow result: Nexus Echo 👋
2024/07/23 19:57:40 Workflow result: ¡Hola! Nexus 👋
```

## Make Nexus calls across Namespaces in Temporal Cloud {#nexus-calls-across-namespaces-temporal-cloud}

This section assumes you are already familiar with [how connect a Worker to Temporal Cloud](https://docs.temporal.io/develop/go/core-application#run-a-temporal-cloud-worker).
The same [source code](https://github.com/temporalio/samples-go/tree/main/nexus) is used in this section, but the `tcld` CLI will be used to create Namespaces and the Nexus Endpoint, and mTLS client certificates will be used to securely connect the caller and handler works to their respective Temporal Cloud Namespaces.

### Install latest tcld CLI and generate certificates

First install the latest version of the `tcld` CLI and generate some certificates for mTLS Worker authentication:

```
brew install temporalio/brew/tcld

> tcld gen ca --org temporal -d 1y --ca-cert ca.pem --ca-key ca.key

Reviewer: ```suggestion
tcld gen ca --org $YOUR_ORG_NAME --validity-period 1y --ca-cert ca.pem --ca-key ca.key
```

```

### Create caller and handler Namespaces

Before deploying to Temporal Cloud, it's necessary to create the appropriate Namespaces for both the caller and handler.

```
tcld login

tcld namespace create \
	--namespace <your-caller-namespace> \
	--region us-west-2 \
	--ca-certificate-file 'path/to/your/ca.pem' \
	--retention-days 1

tcld namespace create \
	--namespace <your-target-namespace> \
	--region us-west-2 \
	--ca-certificate-file 'path/to/your/ca.pem' \
	--retention-days 1
```

Alternatively, you can create Namespace via the UI: [https://cloud.temporal.io/Namespaces](https://cloud.temporal.io/Namespaces)

### Create a Nexus Endpoint to route requests from caller to handler

To create a Nexus Endpoint you must have a `Developer` account role or higher, and have `NamespaceAdmin` permission on the `--target-namespace`.

```
tcld nexus endpoint create \

>   --name <myendpoint> \

Reviewer: Probably want to use the endpoint name from the sample here.

### Run Workers Connected to Temporal Cloud

[View the source code](https://github.com/temporalio/samples-go/tree/main/nexus) in the context of the rest of the application code.

Run the handler Worker:

```go 
cd handler

go run ./worker \
	-target-host <your-target-namespace.account>.tmprl.cloud:7233 \
	-namespace <your-target-namespace.account> \
	-client-cert 'path/to/your/ca.pem' \
	-client-key 'path/to/your/ca.key'
```

Run the caller Worker:

```go 
cd caller

go run ./worker \
	-target-host <your-caller-namespace.account>.tmprl.cloud:7233 \
	-namespace <your-caller-namespace.account> \
	-client-cert 'path/to/your/ca.pem' \
	-client-key 'path/to/your/ca.key'
```

### Start a caller Workflow

In order to start the caller Workflow, run the starter.

```go 
cd caller

go run ./starter \
	-target-host <your-caller-namespace.account>.tmprl.cloud:7233 \
	-namespace <your-caller-namespace.account> \
	-client-cert 'path/to/your/ca.pem' \
	-client-key 'path/to/your/ca.key'
```

This will result in:

```
2024/07/23 19:57:40 Workflow result: Nexus Echo 👋
2024/07/23 19:57:40 Workflow result: ¡Hola! Nexus 👋
```

> ### Observability in Temporal Cloud

Reviewer: Observability is not limited to Temporal Cloud. The only different between the self-hosted and Cloud version is that Cloud supports bidirectional links, and OSS still hasn't caught up due to release timing.

I'd try to clarify this a bit. Maybe put observability in the local section instead of here?

Author: Thanks for the explanation! I changed the titles to more general ones.

	"go.temporal.io/sdk/temporalnexus"
	"go.temporal.io/sdk/workflow"

	"github.com/temporalio/samples-go/nexus/service"
)

// NewSyncOperation is a meant for exposing simple RPC handlers.
var EchoOperation = temporalnexus.NewSyncOperation(service.EchoOperationName, func(ctx context.Context, c client.Client, input service.EchoInput, options nexus.StartOperationOptions) (service.EchoOutput, error) {
	// The method is provided with an SDK client that can be used for arbitrary calls such as signaling, querying,
	// and listing workflows but implementations are free to make arbitrary calls to other services or databases, or
	// perform simple computations such as this one.
	return service.EchoOutput(input), nil
})

```
<!--SNIPEND-->

### Develop an Asynchronous Nexus Operation handler to start a Workflow

Use the 'NewWorkflowRunOperation' constructor, which is the easiest way to expose a Workflow as an operation.
See alternatives [here](https://pkg.go.dev/go.temporal.io/sdk/temporalnexus).

<!--SNIPSTART samples-go-nexus-handler {"selectedLines": ["25-35"]}-->
[nexus/handler/app.go](https://github.com/temporalio/samples-go/blob/main/nexus/handler/app.go)
```go
// ...

> // See alternatives at https://pkg.go.dev/go.temporal.io/sdk/temporalnexus.

Reviewer: This snippet should start either at the start of this comment or strip it out.

The first step in working with Temporal Nexus involves starting a Temporal server with Nexus enabled.

Prerequisites:

- [Install the latest Temporal CLI](https://docs.temporal.io/develop/go/core-application#run-a-development-server)
- [Install the latest Temporal Go SDK](https://docs.temporal.io/develop/go/core-application#install-a-temporal-sdk) (v1.29.1 or higher)

### Start the Temporal Development Server {#start-the-temporal-development-server}

Start the Temporal Development Server by using the server start-dev command, with `system.enableNexus=true`.
The HTTP port is required for Nexus communications.

```
temporal server start-dev --http-port 7243 --dynamic-config-value system.enableNexus=true
```

This command automatically starts the Web UI, creates the default Namespace, and uses an in-memory database.

The Temporal Server should be available on `localhost:7233` and the Temporal Web UI should be accessible at [http://localhost:8233](http://localhost:8233).

## Create caller and handler Namespaces {#create-caller-handler-namespaces}

Before setting up Nexus endpoints, create separate Namespaces for the caller and handler.

```

> temporal operator namespace create --namespace my-target-namespace

Reviewer 2: ```suggestion
temporal operator namespace create --namespace YourNewNamespaceName
```

> temporal operator namespace create --namespace my-caller-namespace

Reviewer 2: ```suggestion
temporal operator namespace create --namespace YourNewCallerNamespaceName
```

```

## Create a Nexus Endpoint to route requests from caller to handler {#create-nexus-endpoint}

After establishing caller and handler Namespaces, the next step is to create a Nexus Endpoint to route requests.

```go
temporal operator nexus endpoint create \

>   --name my-nexus-endpoint-name \

Reviewer 2: ```suggestion
  --name YourNexusEndpointName \
```

>   --target-namespace my-target-namespace \

Reviewer 2: ```suggestion
  --target-namespace YourTargetNamespaceName \
```

>   --target-task-queue my-handler-task-queue

Reviewer 2: ```suggestion
  --target-task-queue YourHandlerTaskQueue
```

```

This will result in:

```
2024/07/23 19:57:40 Workflow result: Nexus Echo 👋
2024/07/23 19:57:40 Started workflow WorkflowID nexus_hello_caller_workflow_20240723195740 RunID c9789128-2fcd-4083-829d-95e43279f6d7
2024/07/23 19:57:40 Workflow result: ¡Hola! Nexus 👋
```

## Make Nexus calls across Namespaces in Temporal Cloud {#nexus-calls-across-namespaces-temporal-cloud}

This section assumes you are already familiar with [how connect a Worker to Temporal Cloud](https://docs.temporal.io/develop/go/core-application#run-a-temporal-cloud-worker).
The same [source code](https://github.com/temporalio/samples-go/tree/main/nexus) is used in this section, but the `tcld` CLI will be used to create Namespaces and the Nexus Endpoint, and mTLS client certificates will be used to securely connect the caller and handler works to their respective Temporal Cloud Namespaces.

### Install the latest tcld CLI and generate certificates

To install the latest version of the `tcld` CLI, run the following command: 
```
brew install temporalio/brew/tcld
```
If you don’t already have certificates, you can generate them for mTLS Worker authentication using the command below:

```

> tcld gen ca --org $YOUR_ORG_NAME --validity-period 1y --ca-cert ca.pem --ca-key ca.key

Reviewer 2: ```suggestion
tcld generate-certificates  certificate-authority-certificate \
    --org $YOUR_ORG_NAME \
    --validity-period 1y \
    --ca-certificate-file /path/to/ca.pem \
    --ca-key-file /path/to/ca.key
```

```

### Create caller and handler Namespaces

Before deploying to Temporal Cloud, ensure the appropriate Namespaces are created for both the caller and handler. 
If you already have these Namespaces, this step is optional.

```
tcld login

tcld namespace create \

> 	--namespace <your-caller-namespace> \

Reviewer 2: ```suggestion
	--namespace <caller_namespace_id>.<account_id> \
```

	--region us-west-2 \
	--ca-certificate-file 'path/to/your/ca.pem' \
	--retention-days 1

tcld namespace create \

> 	--namespace <your-target-namespace> \

Reviewer 2: ```suggestion
	--namespace <target_namespace_id>.<account_id> \
```

```
tcld nexus endpoint create \
  --name <my-nexus-endpoint-name> \
  --target-task-queue my-handler-task-queue \
  --target-namespace <my-target-namespace.account> \
  --allow-namespace <my-caller-namespace.account> \
  --description-file description.md
```

The `--allow-namespace` is used to build an Endpoint allowlist of caller Namespaces that can use the Nexus Endpoint, as described in Runtime Access Control.

Alternatively, you can create a Nexus Endpoint through the UI: [https://cloud.temporal.io/nexus](https://cloud.temporal.io/nexus)
You can also create a Nexus endpoint through the UI in the development server.

### Run Workers Connected to Temporal Cloud

[View the source code](https://github.com/temporalio/samples-go/tree/main/nexus) in the context of the rest of the application code.

Run the handler Worker:

```go
cd handler

go run ./worker \
	-target-host <your-target-namespace.account>.tmprl.cloud:7233 \
	-namespace <your-target-namespace.account> \
	-client-cert 'path/to/your/ca.pem' \
	-client-key 'path/to/your/ca.key'
```

Run the caller Worker:

```go
cd caller

go run ./worker \

> 	-target-host <your-caller-namespace.account>.tmprl.cloud:7233 \

Reviewer 2: ```suggestion
	-target-host <caller_namespace_id>.<account_id>.tmprl.cloud:7233 \
```

> 	-namespace <your-caller-namespace.account> \

Reviewer 2: ```suggestion
	-namespace <caller_namespace_id>.<account_id> \
```

	-client-cert 'path/to/your/ca.pem' \
	-client-key 'path/to/your/ca.key'
```

### Start a caller Workflow

In order to start the caller Workflow, run the starter.

```go
cd caller

go run ./starter \

> 	-target-host <your-caller-namespace.account>.tmprl.cloud:7233 \

Reviewer 2: ```suggestion
	-target-host <caller_namespace_id>.<account_id>.tmprl.cloud:7233 \
```

	-namespace <your-caller-namespace.account> \
	-client-cert 'path/to/your/ca.pem' \
	-client-key 'path/to/your/ca.key'
```

This will result in:

```
2024/07/23 19:57:40 Workflow result: Nexus Echo 👋
2024/07/23 19:57:40 Workflow result: ¡Hola! Nexus 👋
```

## Observability

### Web UI

A synchronous Nexus Operation will surface in the caller Workflow as follows, with just `NexusOperationScheduled` and `NexusOperationCompleted` events in the caller’s Workflow history:

![Observability Sync](/img/nexus-go/go-sdk-observability-sync.png)

An asynchronous Nexus Operation will surface in the caller Workflow as follows, with `NexusOperationScheduled`, `NexusOperationStarted`, and `NexusOperationCompleted`, in the caller’s Workflow history:

![Observability Async](/img/nexus-go/go-sdk-observability-async.png)

### Temporal CLI

Nexus pending operations and callbacks are included in the Workflow description:

```

> temporal workflow describe -w <ID>

Reviewer 2: ```suggestion
temporal workflow describe \
  --workflow-id YourWorkflowID
```

```

Nexus events are included in the caller’s Workflow history:

```

> temporal workflow show -w <ID>

Reviewer 2: ```suggestion
temporal workflow show \
    -workflow-id YourWorkflowID
```

Reviewer: I don't think there's much of a rush to merge this. Up to you if you want to postpone and handle in a separate PR.
